### ISSUE TITLE
cURL line in status report (from Simpletest module) needs more (or less?) info

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#6121]
## Description of the bug

The Simpletest module places a line in the Status Report like this if cURL is installed:

![image](https://github.com/backdrop/backdrop-issues/assets/29214084/a5b25aa7-6f43-4208-95ee-8e0d02ef003e)

This is not terribly informative. If cURL isn't installed, then there is an error message that explains the need, 

`'The testing framework could not be installed because the PHP <a href="@hash_url">hash</a> extension is disabled.'`

but when it is installed, there is no indication of who needs or cares about it.

(This came up in [this Ubercart issue](https://github.com/backdrop-contrib/ubercart/issues/450), which used the same idiom and ended up with up to four identical (and similarly opaque) status lines.)

We could make this line a little bit more useful by adding a `'description'` field that says something like "cURL is required by the Simpletest testing framework," so that people will know where that line of the report is coming from.

Although, in the interest of decluttering the status report, I'd raise the question of whether we even need to list cURL in the status report if it is properly installed. It makes sense to list things like PHP version, MySQL version, etc., but for a simple "either it's there or it's not", maybe "silence on success" would be a good philosophy.

**Comment by klonos** [comment]
Sometimes there are vulnerabilities with certain versions of cUR, and ITSEC teams need to know if their website is patched against them when auditing company assets. So I think that eventually it would be beneficial to move this into system.install and be showing the version of cURL in the status report regardless of whether simpletest is enabled or not (other contrib or custom code in the site might also be using it). The Simpletest module could then hook onto that and add something to the description to denote that it requires it.

For now, we can do something like this in simpletest.install to show the version:
```php
  $curl_found = $curl_version = curl_version() ? $curl_version['version'] . ', ' . $curl_version['ssl_version'] . ', ' . $curl_version['libssh_version'] : $t('Enabled');
  $requirements['curl'] = array(
    'title' => $t('cURL'),
    'value' => $has_curl ? $curl_found : $t('Not found'),
  );
```

As it is now, we cannot override/alter requirements hooks that other modules have set. Once #6241 is done though, we could move this in system.install, and have simpletest.install only append `The testing framework could not be installed because the PHP cURL library is not available.` to the requirement description 😉

**Comment by bugfolder** [comment]
So how about for now if we move the reporting of cURL's version (or absence) into system core (as `$requirements['curl']`), and then in Simpletest, we only add a separate entry `$requirements['simpletest_curl']` if cURL is missing.

Then, when `hook_requirements_alter()` is in, we can combine both bits of information into `$requirements['curl']`.

Other modules that require cURL may still want to create their own `$requirements` entries if they want to maintain compatibility with older versions of core (see https://github.com/backdrop-contrib/ubercart/issues/450).

**Comment by findlabnet** [comment]
For the record, in one of my modules I do check that way:
```
function abuseipdb_report_requirements($phase) {
  $requirements = array();
  $t = get_t();
  $has_curl = function_exists('curl_init');
  
  if ($phase == 'install') {
    if (!$has_curl) {
      $requirements['curl']['severity'] = REQUIREMENT_ERROR;
      $requirements['curl']['description'] = $t('The "AbuseIPDB report" module could not be installed because the PHP <a href="@curl_url">cURL</a> library is not available.', array('@curl_url' => 'http://php.net/manual/en/curl.setup.php'));
    }
  }
  
  return $requirements;
}
```

**Comment by bugfolder** [comment]
@findlabnet, if you're setting the `'description'` of `$requirements['curl']` in `hook_requirements()`, that would wipe out another module's description  (or vice-versa) if they both needed cURL and added their own description for `$requirements['curl']`. But that's another good use case for `hook_requirements_alter()`.

**Comment by findlabnet** [comment]
But in my case only for specific module 'install' phase. 
Anyway, if we're going to standardize something, I'm open to suggesting/changing it.

**Comment by klonos** [comment]
> So how about for now if we move the reporting of cURL's version (or absence) into system core...

Yup, that sounds like a good plan @bugfolder 👍🏼

**Comment by klonos** [comment]
...another nit: the error we are showing is this:
> The testing framework could not be installed because the PHP [cURL](http://php.net/manual/en/curl.setup.php) library is not available.

I don't think that the `could not be installed` wording is accurate here, because the module is actually installed when we show this. Right?

**Comment by kiamlaluno** [comment]
`hook_requirements()` is also invoked when a module is being installed. In that case, the error message is correct; it is not correct when the module is updated, or during runtime.

**Comment by bugfolder** [comment]
PR updated, ready for further testing. Also fixed some capitalization consistency, some "that/which" correction, and the `$t` mentioned above.

**Comment by bugfolder** [comment]
CSpell didn't like "libssh" (3 places), so I added it to .cspell/cspell.json.

**Comment by kiamlaluno** [comment]
It is correct to use the value returned by `get_t()` when `$phase` is equal to `'install'`, which happens when a module is going to be installed.

**Comment by bugfolder** [comment]
> It is correct to use the value returned by get_t() when $phase is equal to 'install', which happens when a module is going to be installed.

That's true, but it's only strictly necessary if the module is going to be installed by `install.php`, before the full Backdrop API is available. Simpletest, though part of core, is not installed (enabled) during a core install; thus, we can rely on `t()` being available when it is installed later. (Which is more or less what @klonos was saying above.) See https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_requirements/1 for details.

**Comment by klonos** [comment]
> Which is more or less what @klonos was saying above.

Yup 👍🏼 

> `hook_requirements()` is also invoked when a module is being installed. ...

But that doesn't prevent the installation AFAIK. Does it?

**Comment by klonos** [comment]
> CSpell didn't like "libssh" (3 places), so I added it to .cspell/cspell.json.

Better to add the word in the custom dictionary though. See my comments in the PR as to why.

**Comment by kiamlaluno** [comment]
If `hook_requirements()` returns an error, the module is not installed.

**Comment by bugfolder** [comment]
Thanks, @kiamlaluno for the clarification and explanation. `$t` is now restored to `simpletest_requirements()`.

**Comment by bugfolder** [comment]
@klonos, your PR comments are (I hope sufficiently) addressed in the revised PR. Check again?

**Comment by klonos** [comment]
@bugfolder I think we may have a deeper core problem here... I've set up your PR branch on my local in order to do some testing, and I tried to mock-trigger those requirement errors. I started by changing the `if (!$has_curl) {` line in `simpletest_requirements()` to the opposite `if ($has_curl) {`, and that threw the following PHP error:
> TypeError: strcasecmp(): Argument `#2` ($string2) must be of type string, array given in strcasecmp() (line 3430 of /app/docroot/core/modules/system/system.module).

Sure enough, after adding a few `dpm()`s here and there, it turns out that the `'curl'` key in the requirements array ends up having arrays instead of strings for its `'title'`/`'value'`/`'description'`:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/0da4d6a3-b1b7-413e-b922-e698e87b55da)
...and that's because of the `array_merge_recursive()` used in `module_invoke_all()`.

I haven't tested, but I'm sure that this is not caused by this PR - it just helped surface the problem: if more than one modules provide the same key in the `$requirements` array, you end up with this situation. We just never tried to do that before.

So we'll need to come up with a solution for that first I guess.

**Comment by klonos** [comment]
There you go: https://www.drupal.org/project/drupal/issues/2865599

**Comment by argiepiano** [comment]
This is a known behavior of `module_invoke_all()`. I had to deal with this problem when working on Entity Tokens - receiving arrays instead of strings when doing a token replacement, because of collision with core tokens that used the same token strings (e.g. `[node:field_my_name]' trying to be replaced both by core tokens and Entity Tokens). After many hours on this, I came to accept that there is no way to work around this problem other than avoiding having multiple hook implementations returning arrays with the same keys (BTW, this is the main reason why in D7, Token uses low dashes, and Entity Tokens uses regular dashes - otherwise you get arrays instead of string replacements). 

I believe this behavior would be difficult if not impossible to change in core, but perhaps there is a way to adapt `update_check_requirements()` to anticipate that arrays may be returned.

**Comment by klonos** [comment]
> ... BTW, this is the main reason why in D7, Token uses low dashes, and Entity Tokens uses regular dashes - otherwise you get arrays instead of string replacements

Oh wow! ...now I know why. Thanks @argiepiano 🙏🏼 

> I believe this behavior would be difficult if not impossible to change in core, but perhaps there is a way to adapt `update_check_requirements()` to anticipate that arrays may be returned.

I agree on both points 👍🏼 ...I've filed #6317 to see if there's something that we can do about it.

**Comment by bugfolder** [comment]
Interesting. Meanwhile, we can here do what I've done with Ubercart's curl checks: use a module-specific key (`simpletest_curl`) in modules that talk about the same topic.

**Comment by bugfolder** [comment]
There are now two entries for "cURL", with the error one from Simpletest at the top. I think that's appropriate.

<img width="1182" alt="image" src="https://github.com/backdrop/backdrop-issues/assets/29214084/1b88e917-c52f-47fe-9ab2-7e7c629aaad5">

**Comment by bugfolder** [comment]
Even without https://github.com/backdrop/backdrop-issues/issues/6241, I think that giving Simpletest its own unique `hook_requirements()` key, as above (and in the current PR) is a perfectly acceptable solution. Yes, there would be two rows of this table that reference cURL (one of them an error notice), but then, even if https://github.com/backdrop/backdrop-issues/issues/6317 were the solution path, there would be two rows that reference cURL; I don't see multiple rows as being a problem.

**Comment by klonos** [comment]
> I don't see multiple rows as being a problem.

I do 😅 ...I actually want to reduce the complexity and amount of rows we have in the status page. See #5863, #5864, #5758.

However, I realize that we currently have many blockers, so getting this 100% right may take some time. So in the interest of not delaying this further, I am fine with having 2 separate rows for cURL, but can we please add a `@todo` (ideally in the Simpletest module) to come and merge those rows later when we have #6241 and/or #6317?

**Comment by bugfolder** [comment]
I"m happy to add the @todo to keep the ball rolling on the PR, but I will note that the extra row only appears when there's a problem, and it appears at the top of the page with a red X indicating that it needs attention. In day-to-day operations, there's no additional row.

**Comment by bugfolder** [comment]
@klonos, can you do a (hopefully final) review of this PR?

**Comment by avpaderno** [comment]
I agree: This is one of the cases where `hook_requirements_alter()` would make sense. With that hook, what described in https://github.com/backdrop/backdrop-issues/issues/6121#issuecomment-1826844300 would be possible.
